test/extensions/filters/http/ext_proc/ext_proc_integration_test.cc

```diff
@@ -1720,4 +1763,148 @@ TEST_P(ExtProcIntegrationTest, PerRouteGrpcService) {
+  upstream_request_->encodeHeaders(Http::TestResponseHeaderMapImpl{{":status", "200"}}, false);
+  upstream_request_->encodeData(100, true);
+  processResponseHeadersMessage(*grpc_upstreams_[0], false,
+                                [](const HttpHeaders&, HeadersResponse&) { return true; });
+  // Verify downstream client receives 200 okay. i.e, no timeout happened.
+  verifyDownstreamResponse(*response, 200);
+}
+
+// Test only the first new timeout message in one state is accepted.
+TEST_P(ExtProcIntegrationTest, RequestMessageNoMutationMultipleNewTimeout) {
+  // Set envoy filter timeout to be 200ms.
+  proto_config_.mutable_message_timeout()->set_nanos(200000000);
+  // Config max_message_timeout proto to 10s to enable the new timeout API.
+  proto_config_.mutable_max_message_timeout()->set_seconds(10);
+  initializeConfig();
+  HttpIntegrationTest::initialize();
+  auto response = sendDownstreamRequest(absl::nullopt);
+
+  processRequestHeadersMessage(*grpc_upstreams_[0], true,
+                               [this](const HttpHeaders&, HeadersResponse&) {
+                                 // Send a 500ms new timeout update first.
+                                 serverSendNewTimeout(500);
+                                 // Server wait for 100ms.
+                                 timeSystem().advanceTimeWaitImpl(100ms);
+                                 // Send the 2nd 10ms new timeout update.
+                                 // The update is ignored by Envoy. No timeout event
+                                 // happened and the traffic went through fine.
+                                 serverSendNewTimeout(10);
+                                 // Server wait for 300ms
+                                 timeSystem().advanceTimeWaitImpl(300ms);
+                                 return true;
+                               });
+
+  ASSERT_TRUE(fake_upstreams_[0]->waitForHttpConnection(*dispatcher_, fake_upstream_connection_));
+  ASSERT_TRUE(fake_upstream_connection_->waitForNewStream(*dispatcher_, upstream_request_));
+  ASSERT_TRUE(upstream_request_->waitForEndStream(*dispatcher_));
+  upstream_request_->encodeHeaders(Http::TestResponseHeaderMapImpl{{":status", "200"}}, false);
+  upstream_request_->encodeData(100, true);
+  processResponseHeadersMessage(*grpc_upstreams_[0], false,
+                                [](const HttpHeaders&, HeadersResponse&) { return true; });
+  verifyDownstreamResponse(*response, 200);
+}
+
+// Setting new timeout 0ms which is < 1ms. The message is ignored.
+TEST_P(ExtProcIntegrationTest, RequestMessageNewTimeoutNegativeTestTimeoutTooSmall) {
+  // Config max_message_timeout proto to 10s to enable the new timeout API.
+  max_message_timeout_ms_ = 10000;
+  newTimeoutWrongConfigTest(0);
+}
+
+// Setting max_message_timeout proto to be 100ms. And send the new timeout 500ms
+// which is > max_message_timeout(100ms). The message is ignored.
```

Comment:
It's hard to tell if the code is actually doing anything either though.

Reply:
Sorry, for the confusion. Let me add some details:

This test is to verify the upper bound check is working. i.e, if the override_message_timeout in the API is greater than the max_message_timeout in the ext_proc filter config, then the message is dropped and timeout happened. Then downstream client receives 500.  All of these logic is in newTimeoutWrongConfigTest().  This test is the counterpart of the previous test, which verifies the lower bound check is working.

In terms of  how this new overrid_message_timeout API works,  the newly added test :RequestMessageNoMutationMultipleNewTimeout is a good example, like the initial timeout is 200ms, and ext_proc server first sends this new API with 500ms new timeout, then stay idle for 300ms, then send back the response. And traffic goes through fine, and downstream client receives 200.  Without the new API, which override the original timer with the new 500ms timer, Envoy ext_proc filter will timeout and send back 500 to the downstream client.

Reply:
OK, got it, thanks.